runtime/builtin_types.go

```diff
@@ -617,6 +617,85 @@ func builtinSorted(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
 	return result, nil
 }
 
+func builtinSum(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	expectedTypes := []*Type{ObjectType, ObjectType}
+	if argc == 1 {
+		expectedTypes = expectedTypes[:1]
+	}
+	if raised := checkFunctionArgs(f, "sum", args, expectedTypes...); raised != nil {
+		return nil, raised
+	}
+	iter, raised := Iter(f, args[0])
+	if raised != nil {
+		return nil, raised
+	}
+	var result *Object
+	if argc > 1 {
+		if args[1].typ == StrType {
+			return nil, f.RaiseType(TypeErrorType, "sum() can't sum strings [use ''.join(seq) instead]")
+		}
+		result = args[1]
+	} else {
+		result = NewInt(0).ToObject()
+	}
+	if result.typ == IntType {
+		item, raised := Next(f, iter)
+		for ; raised == nil; item, raised = Next(f, iter) {
+			if item.typ == IntType {
+				result = NewInt(toIntUnsafe(result).Value() + toIntUnsafe(item).Value()).ToObject()
+				continue
+			}
+			result, raised = Add(f, result, item)
+			break
+		}
+		if raised != nil {
+			if raised.isInstance(StopIterationType) {
+				f.RestoreExc(nil, nil)
+				return result, nil
+			}
+			return nil, raised
+		}
+	}
+	if result.typ == FloatType {
```

Comment:
I see. I think we should prefer simplicity for now instead of optimizing at this state. Especially in light of the additional complexity that will be required to address my comment above about toIntUnsafe.